File: fetcher/fetcher_test.go
package fetcher

import (
	"bytes"
	"io"
	"net/http"
	"testing"

	"github.com/google/go-cmp/cmp"

	"github.com/yagihash/fsw-calendar/fetcher/class"
	"github.com/yagihash/fsw-calendar/fetcher/course"
)

func TestNew(t *testing.T) {
	New("example.com", course.RC, class.S4, http.DefaultClient)
}

func TestFetcher_FetchDocEvents(t *testing.T) {
	t.Run("success to fetch", func(t *testing.T) {
		c := NewTestClient(func(req *http.Request) *http.Response {
			switch req.URL.Path {
			case "/driving/sports/rc/t-4/2023/11.html":
				return &http.Response{
					StatusCode: http.StatusOK,
					Body:       io.NopCloser(bytes.NewBuffer([]byte("<table id=\"table-calendar\"><tbody><tr id=\"2023-11-2\" class=\"row-rc is-show\" data-date=\"2023-11-02\"><td class=\"day this-month thu\" data-date=\"2023/11/02\"><div class=\"inner-table-cell\">2<span class=\"yobi\">（木<span class=\"holiday-text\"></span>）</span></div></td><td class=\"type\"><div class=\"rc\"><p class=\"cell\">T-4 X</p></div></td><td class=\"time\"><div class=\"rc\"><p class=\"cell\">15:20~15:50</p></div></td><td class=\"wait\"><div class=\"rc\"><p class=\"cell\">30分</p></div></td></tr></tbody></table>"))),
					Header:     make(http.Header),
				}
			case "/driving/sports/rc/t-4/2023/12.html":
				return &http.Response{
					StatusCode: http.StatusOK,
					Body:       io.NopCloser(bytes.NewBuffer([]byte("<table id=\"table-calendar\"><tbody><tr id=\"2023-12-1\" class=\"row-rc is-show\" data-date=\"2023-12-01\"><td class=\"day this-month fri\" data-date=\"2023/12/01\"><div class=\"inner-table-cell\">2<span class=\"yobi\">（金<span class=\"holiday-text\"></span>）</span></div></td><td class=\"type\"><div class=\"rc\"><p class=\"cell\">T-4 X</p></div></td><td class=\"time\"><div class=\"rc\"><p class=\"cell\">15:20~15:50</p></div></td><td class=\"wait\"><div class=\"rc\"><p class=\"cell\">30分</p></div></td></tr></tbody></table>"))),
					Header:     make(http.Header),
				}
			default:
				return &http.Response{
					StatusCode: http.StatusOK,
					Body:       io.NopCloser(bytes.NewBuffer([]byte(""))),
					Header:     make(http.Header),
				}
			}
		})
		f := New("example.com", course.RC, class.T4, c)
		got, err := f.FetchDocEvents(2023, 11, 2)
		want := []DocEvent{{Date: "2023-11-02", Start: "15:20", End: "15:50", Title: "T-4 X"}, {Date: "2023-12-01", Start: "15:20", End: "15:50", Title: "T-4 X"}}

		if err != nil {
			t.Errorf("got an unexpexted error: %v", err)
		}

		if diff := cmp.Diff(got, want); diff != "" {
			t.Errorf("got unexpected diff:\n%s", diff)
		}
	})

	t.Run("error http not found", func(t *testing.T) {
		c := NewTestClient(func(req *http.Request) *http.Response {
			return &http.Response{
				StatusCode: http.StatusNotFound,
				Body:       io.NopCloser(bytes.NewBuffer([]byte(""))),
				Header:     make(http.Header),
			}
		})
		f := New("example.com", course.RC, class.T4, c)
		_, err := f.FetchDocEvents(2023, 11, 1)

		if err.Error() != "got status code 404 on https://example.com/driving/sports/rc/t-4/2023/11.html" {
			t.Errorf("got an unexpected error: %v", err)
		}
	})
}

type RoundTripFunc func(req *http.Request) *http.Response

func (f RoundTripFunc) RoundTrip(req *http.Request) (*http.Response, error) {
	return f(req), nil
}

func NewTestClient(fn RoundTripFunc) *http.Client {
	return &http.Client{
		Transport: fn,
	}
}

File: src/fetcher/fetcher.test.ts
import { describe, it, expect, vi, afterEach } from 'vitest';
import { fetchDocEvents } from './fetcher.js';
import { RC } from './course.js';
import { T4 } from './class.js';

function mockHtml(date: string, title: string, time: string): string {
  return `<table id="table-calendar"><tbody>
    <tr class="row-rc" data-date="${date}">
      <td class="type"><div><p>${title}</p></div></td>
      <td class="time"><div><p>${time}</p></div></td>
    </tr>
  </tbody></table>`;
}

function makeResponse(status: number, body: string): Response {
  return { status, text: () => Promise.resolve(body) } as unknown as Response;
}

afterEach(() => {
  vi.unstubAllGlobals();
});

describe('fetchDocEvents', () => {
  it('fetches events across multiple months', async () => {
    vi.stubGlobal('fetch', (url: string) => {
      if ((url as string).includes('/2023/11.html')) {
        return Promise.resolve(makeResponse(200, mockHtml('2023-11-02', 'T-4 X', '15:20~15:50')));
      }
      return Promise.resolve(makeResponse(200, mockHtml('2023-12-01', 'T-4 X', '15:20~15:50')));
    });

    const got = await fetchDocEvents('example.com', RC, T4, 2023, 11, 2);
    expect(got).toEqual([
      { date: '2023-11-02', start: '15:20', end: '15:50', title: 'T-4 X' },
      { date: '2023-12-01', start: '15:20', end: '15:50', title: 'T-4 X' },
    ]);
  });

  it('throws on 404 for the first month', async () => {
    vi.stubGlobal('fetch', () => Promise.resolve(makeResponse(404, '')));

    await expect(fetchDocEvents('example.com', RC, T4, 2023, 11, 1))
      .rejects.toThrow('got status code 404 on https://example.com/driving/sports/rc/t-4/2023/11.html');
  });

  it('stops gracefully on 404 for subsequent months', async () => {
    vi.stubGlobal('fetch', (url: string) => {
      if ((url as string).includes('/2023/11.html')) {
        return Promise.resolve(makeResponse(200, mockHtml('2023-11-02', 'T-4 X', '15:20~15:50')));
      }
      return Promise.resolve(makeResponse(404, ''));
    });

    const got = await fetchDocEvents('example.com', RC, T4, 2023, 11, 2);
    expect(got).toEqual([
      { date: '2023-11-02', start: '15:20', end: '15:50', title: 'T-4 X' },
    ]);
  });

  it('returns empty array for page with no events', async () => {
    vi.stubGlobal('fetch', () => Promise.resolve(makeResponse(200, '<html></html>')));

    const got = await fetchDocEvents('example.com', RC, T4, 2023, 11, 1);
    expect(got).toEqual([]);
  });

  it('skips rows where time has no tilde separator', async () => {
    const html = `<table id="table-calendar"><tbody>
      <tr class="row-rc" data-date="2023-11-02">
        <td class="type"><div><p>T-4 X</p></div></td>
        <td class="time"><div><p>終了</p></div></td>
      </tr>
    </tbody></table>`;
    vi.stubGlobal('fetch', () => Promise.resolve(makeResponse(200, html)));

    const got = await fetchDocEvents('example.com', RC, T4, 2023, 11, 1);
    expect(got).toEqual([]);
  });
});
